Support Swift 5.0

sgoodwin opened the pull request:
This adds support for Swift 5.0 and fixes a bug with the iSO6081 date formatter.

sgoodwin committed:
Fix for JSON decoding with ISO6081 date format.

The normal date formatter doesn't handle perfectly valid ISO6081 dates
in all cases, so I need to put in this custom decoder to handle it
properly.

sgoodwin committed:
Merge branch 'master' of https://github.com/thoughtbot/Swish

sgoodwin committed:
Update Swish for Swift 5.0.

Using the latest Quick and Nimble was pretty much the only code change
required.

We no longer need a framework for Result (even though this is was a good
one). It's included in the standard library now, so the only thing we
needed was an extension to add value and error to keep the tests happy
(and anyone who was depending on those in their code).

gfontenot commented on Source/Protocols/Request.swift:
```diff
@@ -12,6 +11,8 @@ public protocol Request {
   func parse(_ data: Data) throws -> ResponseObject {
+    let decoder = JSONDecoder()
+    decoder.dateDecodingStrategy = .formatted(DateFormatter.iso8601Full)
     return try JSONDecoder().decode(ResponseObject.self, from: data)
```
Should this be using the `decoder` created above?

gfontenot commented on Swish.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>IDEDidComputeMac32BitWarning</key>
```
I think we want to ignore this file

gfontenot reviewed (approved):
Do we also need to make fixes for Swift 5 SPM?

sgoodwin committed:
Adjust from feedback.

Needed to ignore a new file and I was incorrectly creating a new decoder
in my fix (not sure how this worked in my tests like this).

sgoodwin commented:
We don't currently support SPM, do we?

sgoodwin commented:
Hey, wanna point out for the record: I can't merge this.

sgoodwin committed:
Removing the need for Result.

Now that Swift 5 provides this, we don't need it.

sharplet commented on Source/Extensions/Result+Value.swift:
```diff
@@ -0,0 +1,29 @@
+import Foundation
+
+extension Result {
+  public var value: Success? {
```
I don’t think Swish should be extending the standard library’s `Result` here. Given that this is already a breaking change, I think it would be fine to remove this extension.

sharplet commented on Source/Protocols/Request.swift:
```diff
@@ -12,11 +11,24 @@ public protocol Request {
   func parse(_ data: Data) throws -> ResponseObject {
-    return try JSONDecoder().decode(ResponseObject.self, from: data)
+    let decoder = JSONDecoder()
+    decoder.dateDecodingStrategy = .formatted(DateFormatter.iso8601Full)
```
I can see the value in improving our default handling of ISO 8601 date formats, as well as other `JSONDecoder` configuration options. However, I think this could use some more discussion, and would prefer not to tie this to a Swift 5 update. Would you mind separating this change out into a separate PR, or an issue so we can discuss alternative designs?

Have you tried overriding this implementation of `parse(_:)` in your project? Did you run into any issues?

sgoodwin replied:
Sure, I'll move it to a different PR. I have added this to implementations of `parse(_:)`, but for most project this means overriding every single implementation as the default settings don't properly handle ISO8601 like they're supposed to.

sharplet commented on .gitignore:
```diff
@@ -1,6 +1,9 @@
 .DS_Store
 
+# Xcode workspace checks
+IDEWorkspaceChecks.plist
```
I think it’s fine to check this file in rather than ignoring it.

sharplet reviewed (changes requested):
Sorry I didn’t see this sooner — thanks for the PR! I’ve left a few suggestions, but I’d love to get this merged soon.

sgoodwin committed:
Eliminate Result extension.

This caused the compiler to crash initially because there's so much type
inferrence going on in the tests that it couldn't even report errors. I
think I got it worked out to keep the intent behind the test without
having the `value` and `error` fields.

sgoodwin committed:
Remove this for another PR.

So that supporting Swift 5 isn't tied up with including this.

sharplet commented on Tests/Tests/APIClientSpec.swift:
```diff
@@ -33,7 +32,14 @@ class APIClientSpec: QuickSpec {
+                  return nil
+                }
+                
+                expect(value(result)).toEventually(equal("hello world"))
```
I believe this might work as `try? result?.get()`, or even more preferably `try result?.get()` if `expect()` will accept a throwing autoclosure. If the autoclosure form doesn’t work, then an explicit closure definitely should, e.g.: `expect { try result?.get() }`.

sharplet commented on Tests/Tests/APIClientSpec.swift:
```diff
@@ -66,8 +74,15 @@ class APIClientSpec: QuickSpec {
+                }
 
-                expect(result?.value).toEventually(beVoid())
+                expect(value(result)).toEventually(beVoid())
```
Same note here about the shorthand using `result?.get()`.

sharplet commented on Tests/Tests/APIClientSpec.swift:
```diff
@@ -83,7 +98,14 @@ class APIClientSpec: QuickSpec {
+                  return nil
+                }
+                
+                expect(value(result)).toEventually(beVoid())
```
Same note here about the shorthand using `result?.get()`

sharplet commented on Tests/Tests/APIClientSpec.swift:
```diff
@@ -121,8 +143,10 @@ class APIClientSpec: QuickSpec {
-          client.perform(request) {
-            error = $0.error
+          client.perform(request) { result in
+            if case let .failure(resultError) = result {
```
I think this pattern could also be tidied up with `result.get()`:

```swift
var result: Result<Foo, SwishError>?

client.perform(request) {
  result = $0
}

expect { try result?.get() }.toEventually(throwError(SwishError.serverError(code: expectedCode, data: performer.data)))
```

sgoodwin replied:
This one does not appear to work. The function given to expect is expected to produce a `String?`, so the `Predicate` fails to line up type-wise.

sharplet replied:
Ah sorry I was using `Foo` as an example because I didn’t know what the actual type would be. This compiles for me:

```swift
          var result: Result<String, SwishError>?

          client.perform(request) {
            result = $0
          }

          expect { try result?.get() }.toEventually(throwError(SwishError.serverError(code: expectedCode, data: performer.data)))
```

sgoodwin replied:
Ohh I see, the detail is you can expect to throw an error, not just expect to get an error our.

sharplet commented on Tests/Tests/NetworkRequestPerformerSpec.swift:
```diff
@@ -38,8 +37,10 @@ class RequestPerformerSpec: QuickSpec {
-              returnedData = result.value!.data
+              if case let .success(value) = result {
+                returnedCode = value.code
+                returnedData = value.data
```
What do you think about this?
```swift
let value = try? result.get()
returnedCode = value?.code
returnedData = value?.data
```

sharplet commented on Source/Protocols/Request.swift:
```diff
@@ -12,7 +11,8 @@ public protocol Request {
   func parse(_ data: Data) throws -> ResponseObject {
-    return try JSONDecoder().decode(ResponseObject.self, from: data)
+    let decoder = JSONDecoder()
+    return try decoder.decode(ResponseObject.self, from: data)
```
Would you mind inlining the `decoder` variable so that this no longer shows up in the diff?

sharplet reviewed (changes requested):
Thanks a bunch for updating, just a few more tweaks and then I think this will be ready to merge!

sidraval reviewed (approved):
Thanks @sgoodwin !
